x-pack/plugins/apm/server/lib/services/annotations/get_derived_service_annotations.ts

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { isNumber } from 'lodash';
+import { Annotation, AnnotationType } from '../../../../common/annotations';
+import { SetupTimeRange, Setup } from '../../helpers/setup_request';
+import { ESFilter } from '../../../../typings/elasticsearch';
+import { rangeFilter } from '../../helpers/range_filter';
+import {
+  PROCESSOR_EVENT,
+  SERVICE_NAME,
+  SERVICE_ENVIRONMENT,
+  SERVICE_VERSION
+} from '../../../../common/elasticsearch_fieldnames';
+
+export async function getDerivedServiceAnnotations({
+  setup,
+  serviceName,
+  environment
+}: {
+  serviceName: string;
+  environment?: string;
+  setup: Setup & SetupTimeRange;
+}) {
+  const { start, end, client, indices } = setup;
+
+  const filter: ESFilter[] = [
+    { term: { [PROCESSOR_EVENT]: 'transaction' } },
+    { range: rangeFilter(start, end) },
+    { term: { [SERVICE_NAME]: serviceName } }
+  ];
+
+  if (environment) {
+    filter.push({ term: { [SERVICE_ENVIRONMENT]: environment } });
+  }
+
+  const versions =
+    (
+      await client.search({
+        index: indices['apm_oss.transactionIndices'],
+        body: {
+          size: 0,
+          track_total_hits: false,
+          query: {
+            bool: {
+              filter
+            }
+          },
+          aggs: {
+            versions: {
+              terms: {
+                field: SERVICE_VERSION
+              }
+            }
+          }
+        }
+      })
+    ).aggregations?.versions.buckets.map(bucket => bucket.key) ?? [];
+
+  if (versions.length > 1) {
+    const annotations = await Promise.all(
+      versions.map(async version => {
+        const response = await client.search({
+          index: indices['apm_oss.transactionIndices'],
+          body: {
+            size: 0,
+            query: {
+              bool: {
+                filter: filter
+                  .filter(esFilter => !Object.keys(esFilter).includes('range'))
+                  .concat({
+                    term: {
+                      [SERVICE_VERSION]: version
+                    }
+                  })
```

Comment:
I think this would be a lot simpler to read:

```js
[
    { term: { [PROCESSOR_EVENT]: 'transaction' } },
    { term: { [SERVICE_NAME]: serviceName } },
    { term: { [SERVICE_VERSION]: version } }
]
```

Reply:
I've now also added slightly more complicated handling of the environment filter. I'll simplify this to two concats, LMK what you think.